grype/presenter/cyclonedx/presenter.go

```diff
@@ -0,0 +1,85 @@
+package cyclonedx
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+
+	"github.com/anchore/grype/grype/result"
+	"github.com/anchore/grype/grype/vulnerability"
+	"github.com/anchore/syft/syft/pkg"
+	syftCDX "github.com/anchore/syft/syft/presenter/cyclonedx"
+	"github.com/anchore/syft/syft/scope"
+)
+
+// Presenter writes a CycloneDX report from the given Catalog and Scope contents
+type Presenter struct {
+	results          result.Result
+	catalog          *pkg.Catalog
+	scope            scope.Scope
+	metadataProvider vulnerability.MetadataProvider
+}
+
+// NewPresenter is a *Presenter constructor
+func NewPresenter(results result.Result, catalog *pkg.Catalog, theScope scope.Scope, metadataProvider vulnerability.MetadataProvider) *Presenter {
+	return &Presenter{
+		results:          results,
+		catalog:          catalog,
+		metadataProvider: metadataProvider,
+		scope:            theScope,
+	}
+}
+
+// Present creates a CycloneDX-based reporting
+func (pres *Presenter) Present(output io.Writer) error {
+	bom := NewDocumentFromCatalog(pres.catalog, pres.results, pres.metadataProvider)
+
+	srcObj := pres.scope.Source()
+
+	switch src := srcObj.(type) {
+	case scope.DirSource:
+		bom.BomDescriptor.Component = &syftCDX.BdComponent{
+			Component: syftCDX.Component{
+				Type:    "file",
+				Name:    src.Path,
+				Version: "",
+			},
+		}
+	case scope.ImageSource:
+		var imageID string
+		var versionStr string
+		if len(src.Img.Metadata.Tags) > 0 {
+			imageID = src.Img.Metadata.Tags[0].Context().Name()
+			versionStr = src.Img.Metadata.Tags[0].TagStr()
+		} else {
+			imageID = src.Img.Metadata.Digest
+		}
+		src.Img.Metadata.Tags[0].TagStr()
+		bom.BomDescriptor.Component = &syftCDX.BdComponent{
+			Component: syftCDX.Component{
+				Type:    "container",
+				Name:    imageID,
+				Version: versionStr,
+			},
+		}
+	default:
+		return fmt.Errorf("unsupported source: %T", src)
+	}
+
+	xmlOut, err := xml.MarshalIndent(bom, " ", "  ")
+	if err != nil {
+		return err
+	}
+
+	_, err = output.Write([]byte(xml.Header))
+	if err != nil {
+		return err
+	}
+	_, err = output.Write(xmlOut)
+	if err != nil {
+		return err
+	}
+
+	_, err = output.Write([]byte("\n"))
```

Comment:
Not blocking / for consideration — I'm wondering if it'd be more idiomatic to Go, and possibly less verbose, to use an **encoder** for this, given that our objective is to write output to an `io.Writer`. It seems to me that the `Marshal` functions are for when you want to store bytes (return a `[]byte`), and the `Encode` methods are for when you want to _write_ bytes (to a `Writer`). (Same pattern as the `json` std lib package.)

For reference: https://golang.org/pkg/encoding/xml/#NewEncoder

@wagoodman Want to make sure you see this, too.

Reply:
good call, I can head back to syft an update there